articles/bloom-filters-data-structure/index.md

```diff
@@ -0,0 +1,101 @@
+
+There is a significant reduction in time complexity by making use of memory. We have taken a top-down approach, considering the application first. Let us look at the theoretical aspect of hashing and how bloom filters make use of them.
+
+### Hashing: Theory
+Hashing, as we discussed earlier, was developed to optimally store and retrieve data. Listed below are various operations that can be performed on the hash table Abstract Data Type(ADT).
+
+1. Create Hash Table
+2. Search Hash Table
+3. Insert an element into the table
+4. Delete an element from the table
+5. Delete the hash table
+
+### Hash Table
+The hash table is derived from the array. In the case of an array, we have indices that start from either 0 or 1. These indices are the keys and the value stored at these places are the values. Similarly, hash tables work on key-value pairs, where our keys are custom-defined. For example, dictionaries in python are a good example of hash tables.
+
+A hash table is a generalization of the array. With an array, we store the element whose key is k at a position k of the array. That means, given a key k, we find the element whose key is k by just looking in the k position of the array. This is called direct addressing.
+
+The way we index the values determines the position of the value for a given key. If we directly search for the kth element by specifying its index k, again that is called direct addressing. This is feasible in situations when memory is not a constraint. What if we need extra memory? That is, the number of keys is greater than the number of memory locations possible. In such cases, the flexibility of hash tables can be used.
+
+### Hash Function
+The value is allotted to keys based on the hash function. The hash function can be any function. The only requirement is it allows a new and unique key for different values. If two values get the same key, such a situation is called a collision.
+The hash function is used to transform the key into the index. Obtaining such ideal hash functions is a tough job.
+
+### Characteristics of Good Hash Functions
+An ideal hash function obeys the following properties:
+
+- Minimal number of collisions
+- Not computationally intensive
+- Uniform distribution of keys for various values
+- Maximum usage of the information present in the values
+
+Let us discuss some of the strategies that complement hash functions and deal with collisions.
+
+#### Separate Chaining
+We use the concept of linked lists to resolve collisions. Upon collision for a given key, the new value is linked to the existing value at the location using linked lists. This method is easy to implement but is memory intensive. Each link adds on a new pointer and therefore additional memory space is required. While using the hash table, if the length of the linked list is too long, then the lookup time increases to O(n) from O(1). This defeats the purpose of implementing a hash table.
```

Comment:
put a link/source about linked list?
Also ? is separate chaining and linked lists the same thing?

Reply:
No... separate chaining uses linked list... have clarified the same... Let me know if it is alright